tfhe/src/core_crypto/entities/modulus_switched_lwe_ciphertext.rs

```diff
@@ -0,0 +1,286 @@
+
+        Self {
+            packed_coeffs,
+            lwe_dimension: ct.lwe_size().to_lwe_dimension(),
+            log_modulus,
+            uncompressed_ciphertext_modulus,
+        }
+    }
+
+    /// Converts back a compressed ciphertext to its initial modulus
+    /// The noise added during the compression says int hte output
+    /// The output must got through a PBS to reduce he noise
+    pub fn extract(&self) -> LweCiphertextOwned<Scalar> {
+        let log_modulus = self.log_modulus.0;
+
+        // log_modulus lowest bits set to 1
+        let mask = Scalar::MAX >> (Scalar::BITS - log_modulus);
+
+        let container = (0..(self.lwe_dimension.0 + 1))
+            .map(|i| {
+                let start = i * log_modulus;
+                let end = (i + 1) * log_modulus;
+
+                let start_block = start / Scalar::BITS;
+                let start_remainder = start % Scalar::BITS;
+
+                let end_block_inclusive = (end - 1) / Scalar::BITS;
+
+                if start_block == end_block_inclusive {
+                    //                                  Scalar::BITS
+                    //                                |---------------|
+                    //
+                    // packed_coeffs:  start_block-1  |  start_block  | start_block+1
+                    // container    :             |  i-1  |   i   |  i+1  |
+                    //
+                    //                                    |-------|
+                    //                                   log_modulus2
+                    //
+                    //                                 |--|
+                    //                          start_remainder
+                    //
+                    // container[i] =
+                    self.packed_coeffs[start_block] >> start_remainder
+                } else {
+                    //                                   Scalar::BITS
+                    //                                 |---------------|
+                    // packed_coeffs:  |  start_block  | start_block+1 | start_block+2
+                    // container    :         i-1  |   i   |  i+1  |  i+2  |
+                    //                             |-------|
+                    //                            log_modulus
+                    //
+                    //                 |-----------|
+                    //                start_remainder
+                    //
+                    //                             |---|
+                    //                   Scalar::BITS - start_remainder
+                    //
+                    // container[i] = first_part | second_part
+                    assert_eq!(end_block_inclusive, start_block + 1);
+
+                    let first_part = self.packed_coeffs[start_block] >> start_remainder;
+
+                    let second_part = self.packed_coeffs[end_block_inclusive]
+                        << ((Scalar::BITS - start_remainder) % Scalar::BITS);
```

Comment:
I'm surprised the % Scalar::BITS is required ?

Reply:
It's not